File: system-test/typescript/src/testkit/driver/adapter/api/SystemErrorMapper.ts
import type { SystemError } from '../../port/dtos/errors/SystemError.js';
import type { ProblemDetailResponse } from './client/dtos/errors/ProblemDetailResponse.js';

export class SystemErrorMapper {
  private constructor() {
    // Utility class
  }

  static from(problemDetail: ProblemDetailResponse): SystemError {
    const message = problemDetail.detail ?? 'Request failed';
    const errors = problemDetail.errors ?? [];
    return {
      message,
      fieldErrors: errors.map((e) => ({
        field: e.field,
        message: e.message,
      })),
    };
  }
}

File: system-test/dotnet/Driver.Adapter/Api/SystemErrorMapper.cs
using Driver.Adapter.Api.Client.Dtos.Errors;
using Driver.Port.Dtos.Error;

namespace Driver.Adapter.Api;

public static class SystemErrorMapper
{
    public static SystemError From(ProblemDetailResponse problemDetail)
    {
        var message = problemDetail.Detail ?? "Request failed";

        if (problemDetail.Errors != null && problemDetail.Errors.Count > 0)
        {
            var fieldErrors = problemDetail.Errors
                .Select(e => new SystemError.FieldError(e.Field ?? string.Empty, e.Message ?? string.Empty, e.Code))
                .ToList()
                .AsReadOnly();
            return SystemError.Of(message, fieldErrors);
        }

        return SystemError.Of(message);
    }
}
